### PR TITLE
Styles Navigation Screen: Add Style Book

### LINKED ISSUE #50393: Expose Style Book in Styles details view
Let's add a control to easily dive into the Style Book from the Site Editor > Styles panel. 

## Visual 

CleanShot 2023-05-05 at 18 23 08
Added by @richtabor for feedback.

### PR DESCRIPTION
**Pull request by andrewserong**
## What?

Fixes: https://github.com/WordPress/gutenberg/issues/50393

Add a button to the Styles left-hand navigation screen in the site editor Browse mode. Clicking the button opens the Style Book within the frame, displayed without tabs. Clicking anywhere within the Style Book frame opens up the edit mode with the Style Book open.

## Why?

As raised in #50393, the goal is to make it easier for folks to open up the Style Book and see how styles look across blocks at a glance.

## How?

* Add an extra button to the Styles navigation screen within the browse mode of the site editor
* Use the same icon as the Style Book icon in the global styles sidebar
* Allow the Style Book to be rendered within the frame in browse mode
* Hide the tabs, and treat the Style Book as one big button when displayed in this screen

## Testing Instructions

1. Open up the site editor, and go to the Styles screen on the left hand side
2. Click the eye icon expose the Style Book
3. You should be able to scroll up and down the Style Book. Clicking within it will open up the edit mode / navigate to the full site editor.
4. Navigating from the Styles screen back to the main navigation screen and back to Styles screen should reset the display state for the Style Book.
5. Double-check that otherwise opening/closing the site editor works just as before, and that the Style Book within the site editor is otherwise unaffected.

### Testing Instructions for Keyboard

Similar to above, but tab over to the eye icon from the Styles screen. Also, try tabbing all the way to the Style Book frame. You should be able to press Enter / Space to open up the edit mode.

## Screenshots or screencast 

image

https://github.com/WordPress/gutenberg/assets/14988353/e32f86bd-a7f2-4598-844c-4aeea9bc399b

### CODE REVIEW

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (andrewserong) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: I've just used a simple div here so that there isn't any space between the buttons. Happy to change this if we need the buttons to have some breathing room.

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: I see why you did it this way, and it works well. 👍 

I think dispatches return a promise so you might not even need `useState`

```jsx
onClick={ async () => {
	await setCanvasMode( 'edit' );
	await openGeneralSidebar(
        	'edit-site/global-styles'
	);
	setEditorCanvasContainerView( 'style-book' )
} }
```

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Since we're reusing this, you could chain these up:

```js
	const openGlobalStyles = async () =>
		Promise.all( [
			setCanvasMode( 'edit' ),
			openGeneralSidebar( 'edit-site/global-styles' ),
		] );

...

					<SidebarButton
						icon={ edit }
						label={ __( 'Edit styles' ) }
						onClick={ async () => await openGlobalStyles() }
					/>
```

Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Then, above:

```jsx
					<SidebarButton
						icon={ seen }
						label={ __( 'Style Book' ) }
						onClick={ async () => {
							await openGlobalStyles();
							setEditorCanvasContainerView( 'style-book' );
						} }
					/>
```

Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Seems to work, I think 🤣

**Comment by andrewserong** [comment]
Oh, thanks for the good suggestion @ramonjd, I'll have a play! 🙇

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Sounds good 👍 

I'm not an expert here, but would something like `<div role="group" aria-label="Styles commands">` be appropriate too?

Asking for a friend :trollface:

Comment (andrewserong) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Good question — I only added the wrapping `div` for styling purposes rather than to flag anything semantically. Since there's only two controls at the moment, I wasn't sure how valuable having a group role would be for those two buttons?

Happy to add it in, though, if you think it's better to treat them as a group!

Comment (ramonjd) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Honestly I don't know. I think it's fine to leave as is. If we find out a group role is helpful it's easy to add in 👍

Comment (andrewserong) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Cheers, sounds good 👍

**Review by ramonjd — APPROVED** [review]
Tentatively approving if my async suggestion turns to ash.
This is working as described for me. 
Thank you!

**Comment by andrewserong** [comment]
Nicely done @ramonjd, your suggestion of using async functions and `await` and resolving the promises in a function used by both callbacks works nicely and neatens things up 👍. I've gone with that approach and added an extra comment for clarity just in case anyone in the future wonders why we needed to `await` dispatching those actions in order to get it working.

Thanks again, I'd entirely forgotten that dispatches return promises! 🎉

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Sure does! 🍻

**Comment by mtias** [comment]
My thinking was that we would render the style book in the frame but not navigate to the editor. So that you can check different styles quickly against the style book. If you click on the frame, then you'd go into the editor. Can we try that?

**Comment by richtabor** [comment]
> My thinking was that we would render the style book in the frame but not navigate to the editor.

That's an interesting idea. Would you still be able to navigate the tabs, or as soon as you interact it opens the Edit view?

**Comment by mtias** [comment]
No, the tabs wouldn't work in their current design. But I think we need to iterate there anyways and take them out of the frame, fwiw.

**Comment by andrewserong** [comment]
Thanks for the feedback and ideas, folks! I've had a play with seeing if we can render the Style Book in the canvas within browse mode. It seems to work kind of okay, with a few caveats. Here's how it's currently looking:

https://github.com/WordPress/gutenberg/assets/14988353/07cf1e51-d216-4204-a564-8ab5f2c01198

* Renders in the Style Book in the frame, and tabs can be used to browse through all the blocks (I quite like including the tabs, personally)
* The close button in the top-right works to close the style book, in addition to the eye icon
* Clicking on any of the block previews within the Style Book will open the site editor proper
* Caveat: With this approach we cannot redirect the user to the particular block they clicked on since TabPanel is not yet a controlled component (see: https://github.com/WordPress/gutenberg/pull/46704)
* Caveat: Hovering over the frame still causes it to expand/scale up in the Style Book view. If we want the style book to be interactive / navigable via tabs in this view, then we probably want to prevent the expansion if the style book is open.

Let me know what everyone thinks — if this seems like a good direction, I'm happy to keep tweaking / refining.

At this stage, I'd probably lean toward seeing how we can use the Style Book pretty much as-is within this PR. If we're looking at redesigning the Style Book / removing the tabs, then we might want to split out that work into a separate PR.

**Comment by jasmussen** [comment]
Catching up here, looks good. To Matías note, we need to explore a different interface than the tabs. That can happen separately. But the tabs being functional in the frame, also the close button, that isn't a great experience. 

Can we make it so a click anywhere on the frame, when the style book is open, goes to the fullscreen editor with the global styles panel, and the style book open? And can we hide the tabs in the mean time?

**Comment by ramonjd** [comment]
Nice one!

> The close button in the top-right works to close the style book, in addition to the eye icon

Just to note that the close button [can be hidden](https://github.com/WordPress/gutenberg/blob/4af25a14e9d46fbdaca643df22417b068f0894c1/packages/edit-site/src/components/editor-canvas-container/index.js#L96) if that's desired

> Caveat: With this approach we cannot redirect the user to the particular block they clicked on since TabPanel is not yet a controlled component 

My assumption was that the foundational purpose of this screen was to allow the user to select a style variation on the left hand side. 

I can see the worth of being able to preview different blocks in each style variation, and therefore tab interactions, but clicking on a block and being taken to the site editor might detract from the primary purpose of the page: to select a style variation.

**Comment by andrewserong** [comment]
> Just to note that the close button [can be hidden](https://github.com/WordPress/gutenberg/blob/4af25a14e9d46fbdaca643df22417b068f0894c1/packages/edit-site/src/components/editor-canvas-container/index.js#L96) if that's desired

Oh, thanks!

> Can we make it so a click anywhere on the frame, when the style book is open, goes to the fullscreen editor with the global styles panel, and the style book open? And can we hide the tabs in the mean time?

If we hide the tabs, would we expect to list all blocks out, or just those in the first tab, with the tabs hidden? I'll have a play around and see what's possible.

Thanks again for the feedback and ideas, folks!

File: packages/edit-site/src/components/editor-canvas-container/style.scss
Comment (andrewserong) [inline packages/edit-site/src/components/editor-canvas-container/style.scss:2]: I've removed this rule so that there isn't a tiny bit of white bleeding in at the edge of the Style Book when viewed in browse mode. This was most noticeable on Chrome when hovering over the main frame when the background color of the site was dark. There'd be a thin sliver of the background color of the canvas container poking through. This removal fixes that.

**Comment by andrewserong** [comment]
Update: I think I've managed to get it working pretty well while hiding the tabs in the frame view:

https://github.com/WordPress/gutenberg/assets/14988353/e32f86bd-a7f2-4598-844c-4aeea9bc399b

In the latest update:

* The tabs are hidden in the Style Book in the frame view in Browse mode
* All blocks are currently visible in one big list
* The list is scrollable
* Clicking anywhere within the Style Book opens up the editor mode with the Style Book open
* Users can tab to the Style Book via keyboard and press Enter / Space to open the editor mode
* Clicking back / navigating to the main navigation screen closes the Style Book
* In the editor mode, we still display the tabs as normal

I'm sure there'll be other refinements and tweaks to do, too, but let me know how this feels for balance at the moment.

**Comment by jasmussen** [comment]
Took this for a spin, nice work:

status

It helps a lot that the tabs are omitted, and it gives us the benefit that you can see the various style variations applied across the style book, before saving.

_However_ I just realized, that we might be able to skip this interim step. That is, go directly from this screen:

Screenshot 2023-05-16 at 10 51 25

to this screen:

Screenshot 2023-05-16 at 10 51 30

I should've noted this sooner, sorry, but there are a few benefits to this:

* The eye button behaves like the edit button right next to it, being a deep-link.
* It avoids us having to add a toggled state to it. 
* It makes it just a little bit clearer what the style book is, given the full-screen context.

This also ties into some of the work @beafialho has been doing on the style book, where here's a screen that retires the tabs:

2  Style Book

I recognize the downside that you can't preview the style variations on the style book, and we can potentially revisit this. But it might be a simpler starting point? CC: @jameskoster

**Comment by jameskoster** [comment]
I originally thought that's how this would work (deep link). It looks like that was the initial implementation, but it was later changed on [request](https://github.com/WordPress/gutenberg/pull/50566#issuecomment-1545612652). 

Being able to preview the style variations and view all blocks at the same time seems quite handy to me, as it's not something currently possible in the Styles panel in Edit view. So I don't mind the toggle. Especially as we'll likely have a similar toggle on Template / Part list panels for moving between browse/manage views. 

Until we've done a design exploration that considers Styles to be a more discrete experience I don't have an especially strong opinion on this one.

**Comment by jasmussen** [comment]
Ah, then let's go with this as is! No reason to hold this up.

**Comment by juanfra** [comment]
This looks really nice! It'd be interesting to have some styles for the `is-pressed` buttons on the sidebar navigation to clearly indicate what's been selected, making the "toggle" more explicit. From what I see now, there's only a shadow applied to the focus state of `is-pressed` which can be easily lost. I really love this change; great job!

File: packages/edit-site/src/components/block-editor/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/block-editor/index.js]: Hi @andrewserong 

I have a PR up that allows the consuming component to control the resizable editor:

- https://github.com/WordPress/gutenberg/pull/50682

So style book will be able to pass `enableResizing={ canvasMode !== 'view' }` down to `<EditorCanvasContainer />`. 

I'm not sure whether it makes your life harder or easier 🤣 

But the logic is that, some use cases might want to force "resizeability" either way, e.g., the global styles revisions pane.

Comment (andrewserong) [reply packages/edit-site/src/components/block-editor/index.js]: Oh, thanks Ramon! It won't affect my work too much either way, but I like the approach in that PR. I'll review that shortly, and once it's in I can rebase this PR 🙂

**Comment by andrewserong** [comment]
Thanks for the feedback, folks! I've given this PR a rebase and I believe it should be ready for a final review now.

> It'd be interesting to have some styles for the is-pressed buttons on the sidebar navigation to clearly indicate what's been selected, making the "toggle" more explicit. From what I see now, there's only a shadow applied to the focus state of is-pressed which can be easily lost.

Yes, it'd be good to take a look at the `is-pressed` styles. Since they apply to the `.edit-site-sidebar-button` class in general, I'd be happy to look at this in a follow-up if that works? Do any designers have an opinion on what the pressed state should look like / specific colors to use? I imagine it might look something like a background color of `$gray-200` with a foreground color of `#1e1e1e` to look like the following:

image

> This also ties into some of the work @beafialho has been doing on the style book, where here's a screen that retires the tabs:

Very happy to help efforts to improve the style book — once the designs are ready, shall we open up a separate issue to track that?

> Ah, then let's go with this as is! No reason to hold this up.

Thanks — I can see good arguments between keeping the Style Book within the frame in browse mode (current state of this PR) and opening up edit mode (the original implementation in this PR). It won't be difficult to switch out the approach after merging if we decide we'd like to switch back to opening the full edit mode.

Please let me know if everyone's happy with the current state of this PR, or if you'd like further iterations before landing 🙂

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]:
```suggestion
	const isStyleBookOpened = useSelect(
		( select ) =>
			'style-book' ===
			unlock( select( editSiteStore ) ).getEditorCanvasContainerView(),
		[]
	);
```

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (ramonjd) [inline packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]:
```suggestion
								onClick={ async () =>
									setEditorCanvasContainerView(
										! isStyleBookOpened
											? 'style-book'
											: undefined
									)
								}
```

**Review by ramonjd — APPROVED** [review]
I just left a couple of optional code changes. This is working well for me. 👍 

https://github.com/WordPress/gutenberg/assets/6458278/b475422b-1137-4b1b-a7d3-ef96cf2405df

I can't find any regressions to the style book in `canvas === 'edit'` mode.

I agree it'd be handy for my 🧠 to have an `is-pressed` state on the style book button in the sidebar to indicate that it's open.

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Good call — I noticed that we can also remove `async` so I've committed separately in https://github.com/WordPress/gutenberg/pull/50566/commits/1ddbafe12797efd80c2eb1278e849c357c8da76d

File: packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js
Comment (andrewserong) [reply packages/edit-site/src/components/sidebar-navigation-screen-global-styles/index.js]: Thanks, that's neater — I always wind up doing the long version while hacking around and then forget to shorten! Cheers, committed in https://github.com/WordPress/gutenberg/pull/50566/commits/1ddbafe12797efd80c2eb1278e849c357c8da76d

**Comment by andrewserong** [comment]
Thanks for the re-review and suggestions @ramonjd! I'll leave this PR open overnight to see if there are any other requests from designers, and if not will merge this in tomorrow and/or follow-up with the `is-pressed` state 🙂

**Comment by jasmussen** [comment]
Thanks for all the work. Agreed we should look at the toggled state separately and not let it block this one. We need that designed in a more careful way, as it's something we'll want across other toggle states in the dark material. I feel like @jameskoster may have some thoughts there as well, so I'll defer to him, but just as a baseline, something like this would feel more at home:

Screenshot 2023-05-17 at 09 28 14

I recognize that doesn't check all the boxes, all the more reason to explore it as a followup.

**Review by apeatling — APPROVED** [review]
Great work @andrewserong. This is testing well for me, looks good to merge.

Only thing I noticed was the flickering from white of each style preview box in the sidebar when the styles sidebar first loads. It's a bit jankier than it could be. It'd be good to try and improve that in a follow up.

**Comment by andrewserong** [comment]
Thanks all! I'll merge this in now. Happy to continue tweaking in follow-ups, and please let me know if there were any other ideas you'd like me to look into 🙂

**Pull request merged**
